platform/solr/solr-factory/src/test/groovy/org/codice/solr/factory/impl/HttpSolrClientFactorySpec.groovy

```diff
@@ -0,0 +1,185 @@
+import org.apache.solr.client.solrj.response.SolrPingResponse
+import org.apache.solr.common.util.NamedList
+import org.codice.spock.extension.DeFinalize
+import org.codice.spock.extension.Supplemental
+import org.codice.spock.extension.builtin.DeFinalizer
+import org.junit.runner.RunWith
+import spock.lang.Specification
+import spock.lang.Timeout
+import spock.lang.Unroll
+import spock.util.environment.RestoreSystemProperties
+
+import static java.util.concurrent.TimeUnit.SECONDS
+
+@RestoreSystemProperties
+@Supplemental
+@RunWith(DeFinalizer)
+@DeFinalize(HttpSolrClientFactory)
+class HttpSolrClientFactorySpec extends Specification {
+  static final String CORE = "test_core"
+  static final String DATA_DIR = "data_dir"
+  static final String SOLR_HOST = "server"
+  static final String SOLR_HOST2 = "server2"
+  static final String SOLR_PORT = "1234"
+  static final String SOLR_CONTEXT = "solr"
+  static final String SOLR_URL = "https://$SOLR_HOST:$SOLR_PORT/$SOLR_CONTEXT"
+  static final String CORE_URL = "$SOLR_URL/$CORE"
+  static final String SOLR_URL2 = "https://$SOLR_HOST2:$SOLR_PORT/$SOLR_CONTEXT"
+  static final String CORE_URL2 = "$SOLR_URL2/$CORE"
+
+  static final String CONFIG_XML = "solrconfig.xml"
+  static final String SCHEMA_XML = "schema.xml"
+  static final int AVAILABLE_TIMEOUT_IN_SECS = 25
+
+  def cleanup() {
+    // reset the config store
+    ConfigurationStore.instance.dataDirectoryPath = null
+  }
+
+  @Timeout(HttpSolrClientFactorySpec.AVAILABLE_TIMEOUT_IN_SECS)
+  @Unroll
+  def 'test new client becoming available when system property solr.http url is #solr_url_is'() {
+    given:
+      def httpClient = Mock(SolrClient) {
+        ping() >> Mock(SolrPingResponse) {
+          // verify the Http Solr client should have been successfully pinged at least once
+          // must be done in 'given' because it will be called from a different thread and if
+          // declared in 'then', it will be out of scope and not matched
+          (1.._) * getResponse() >> Mock(NamedList) {
+            get("status") >> "OK"
+          }
+        }
+      }
+      def factory = Spy(HttpSolrClientFactory) {
+        // verify an actual Http Solr client will be created
+        // must be done in 'given' because it will be called from a different thread and if
+        // declared in 'then', it will be out of scope and not matched
+        1 * createSolrHttpClient(solr_url, CORE, core_url) >> httpClient
+      }
+
+    and:
+      System.setPropertyIfNotNull("solr.http.url", solr_http_url)
+      System.setPropertyIfNotNull("org.codice.ddf.system.hostname", system_host)
+      System.setPropertyIfNotNull("org.codice.ddf.system.httpsPort", system_port)
+      System.setPropertyIfNotNull("org.codice.ddf.system.rootContext", system_context)
+
+    and:
+      System.setProperty("solr.data.dir", DATA_DIR)
+
+    when:
+      def client = factory.newClient(CORE)
+
+    then: "the reported core should correspond to the requested one"
```

Comment:
❓Shouldn't we have a separate test to verify that the core matches what was requested instead of doing this here and also in the other test below?

Reply:
Maybe but I was forced to separate the 2 `when`s with a `then` because of the asynchronous nature of the method so it was either doing that test or simply doing `true`.